File: codeforce/pass/f.cpp
#include <iostream>
#include <vector>
#include <cstdlib>

struct Node {
    int size;
    int priority;
    int data;
    bool zero;
    Node* left;
    Node* right;

    Node(int data = 0, bool isZero = true) : data(data), zero(isZero), size(1), left(nullptr), right(nullptr) {
        priority = rand() * rand();
    }
};

class Treap {
private:
    Node* root = nullptr;
    int num = 0;

    int getSize(Node* t) {
        return t ? t->size : 0;
    }

    bool getZero(Node* t) {
        return t ? t->zero : false;
    }

    void updateSize(Node* t) {
        if (t) {
            t->size = getSize(t->left) + getSize(t->right) + 1;
        }
    }

    void updateZero(Node* t) {
        if (t) {
            t->zero = getZero(t->left) || getZero(t->right) || (t->data == 0);
        }
    }

    Node* merge(Node* l, Node* r) {
        if (!l || !r) {
            return l ? l : r;
        }
        if (l->priority > r->priority) {
            l->right = merge(l->right, r);
            updateSize(l);
            updateZero(l);
            return l;
        } else {
            r->left = merge(l, r->left);
            updateSize(r);
            updateZero(r);
            return r;
        }
    }

    void split(Node* t, int key, Node*& l, Node*& r) {
        if (!t) {
            l = r = nullptr;
            return;
        }
        if (getSize(t->left) < key) {
            split(t->right, key - getSize(t->left) - 1, t->right, r);
            l = t;
        } else {
            split(t->left, key, l, t->left);
            r = t;
        }
        updateSize(t);
        updateZero(t);
    }

    Node* searchNull(Node* t, int& ind) {
        Node* cur = t;
        ind = getSize(cur->left);

        while (getZero(cur)) {
            if (cur->left && getZero(cur->left)) {
                cur = cur->left;
                ind -= getSize(cur->right) + 1;
            } else if (getData(cur) == 0) {
                break;
            } else {
                cur = cur->right;
                ind += getSize(cur->left) + 1;
            }
        }

        return cur;
    }

    int getData(Node* t) {
        return t ? t->data : -1;
    }

    void toArray(Node* node, std::vector<int>& a) {
        if (node) {
            toArray(node->left, a);
            a.push_back(getData(node));
            toArray(node->right, a);
        }
    }

    Node* remove(Node* r, int key) {
        Node *t1, *t2, *t3;
        split(r, key, t1, t2);
        split(t2, 1, t2, t3);
        r = merge(t1, t3);
        delete t2;
        return r;
    }

public:
    void build(int n) {
        for (int i = 0; i < n; ++i) {
            root = merge(new Node, root);
        }
    }

    void insert(int ind) {
        Node* l, * r, * z;
        int i;

        split(root, ind, l, r);
        z = searchNull(r, i);
        if (z && z->data == 0) {
            r = remove(r, i);
        }

        root = merge(merge(l, new Node(++num, false)), r);
    }

    void printNonZeroElements() {
        std::vector<int> ans;
        toArray(root, ans);
        while (!ans.empty() && ans.back() == 0) {
            ans.pop_back();
        }

        std::cout << ans.size() << '\n';
        for (int i = 0; i < ans.size(); ++i) {
            std::cout << ans[i] << ' ';
        }
        std::cout << '\n';
    }
};

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, m;
    std::cin >> n >> m;

    Treap tree;
    tree.build(m);

    for (int i = 0; i < n; ++i) {
        int j;
        std::cin >> j;
        tree.insert(j - 1);
    }

    tree.printNonZeroElements();

    return 0;
}

File: codeforce/pass/f.go
package main

import (
	"fmt"
	"math/rand"
)

type Node struct {
	size  int
	p     int
	data  int
	zero  bool
	left  *Node
	right *Node
}

func newNode(data int, isZero bool) *Node {
	return &Node{
		data:  data,
		zero:  isZero,
		p:     rand.Int(),
		size:  1,
		left:  nil,
		right: nil,
	}
}

type Treap struct {
	root *Node
	num  int
}

func (t *Treap) get_size(n *Node) int {
	if n == nil {
		return 0
	}
	return n.size
}

func (t *Treap) get_zero(n *Node) bool {
	if n == nil {
		return false
	}
	return n.zero
}

func (t *Treap) update_size(n *Node) {
	if n == nil {
		return
	}
	n.size = t.get_size(n.left) + t.get_size(n.right) + 1
}

func (t *Treap) update_zero(n *Node) {
	if n == nil {
		return
	}
	n.zero = t.get_zero(n.left) || t.get_zero(n.right) || (n.data == 0)
}

func (t *Treap) get_data(n *Node) int {
	if n == nil {
		return -1
	}
	return n.data
}

func (t *Treap) merge(l *Node, r *Node) *Node {
	if l == nil {
		return r
	}
	if r == nil {
		return l
	}

	if l.p > r.p {
		l.right = t.merge(l.right, r)
		t.update_size(l)
		t.update_zero(l)
		return l
	} else {
		r.left = t.merge(l, r.left)
		t.update_size(r)
		t.update_zero(r)
		return r
	}
}

func (t *Treap) split(n *Node, key int) (*Node, *Node) {
	if n == nil {
		return nil, nil
	}

	var l, r *Node
	if t.get_size(n.left) < key {
		l, n.right = t.split(n.right, key-t.get_size(n.left)-1)
		r = n
	} else {
		n.left, r = t.split(n.left, key)
		l = n
	}
	t.update_size(n)
	t.update_zero(n)
	return l, r
}

func (t *Treap) search_null(n *Node) (*Node, int) {
	if n == nil {
		return nil, 0
	}
	cur := n
	ind := t.get_size(cur.left)

	for t.get_zero(cur) {
		if cur.left != nil && t.get_zero(cur.left) {
			cur = cur.left
			ind -= t.get_size(cur.right) + 1
		} else if t.get_data(cur) == 0 {
			break
		} else {
			cur = cur.right
			ind += t.get_size(cur.left) + 1
		}
		if cur == nil {
			break
		}
	}

	return cur, ind
}

func (t *Treap) to_vector(node *Node, a *[]int) {
	if node != nil {
		t.to_vector(node.left, a)
		*a = append(*a, t.get_data(node))
		t.to_vector(node.right, a)
	}
}

func (t *Treap) remove(n *Node, key int) *Node {
	var t1, t2, t3 *Node
	t1, t2 = t.split(n, key)
	t2, t3 = t.split(t2, 1)
	n = t.merge(t1, t3)
	return n
}

func (t *Treap) build(n int) {
	for i := 0; i < n; i++ {
		t.root = t.merge(newNode(0, true), t.root)
	}
}

func (t *Treap) insert(ind int) {
	var l, r *Node
	l, r = t.split(t.root, ind)
	z, i := t.search_null(r)
	if z != nil && z.data == 0 {
		r = t.remove(r, i)
	}

	t.root = t.merge(t.merge(l, newNode(t.num+1, false)), r)
	t.num++
}

func main() {
	var n, m int
	fmt.Scan(&n, &m)

	tree := Treap{root: nil, num: 0}
	tree.build(m)

	var tmp int
	for i := 0; i < n; i++ {
		fmt.Scan(&tmp)
		tree.insert(tmp - 1)
	}

	var ans []int
	tree.to_vector(tree.root, &ans)
	for len(ans) > 0 && ans[len(ans)-1] == 0 {
		ans = ans[:len(ans)-1]
	}

	fmt.Println(len(ans))
	for _, v := range ans {
		fmt.Printf("%d ", v)
	}
}
